### PR TITLE
[parity-bytes] add no-std support  

### LINKED ISSUE #153: support paritybytes in #![no_std] libraries
In #![no_std] libraries, we need to use `alloc::vec::Vec` instead of  `std::vec::Vec`, like `bitvec` did.
https://github.com/myrrlyn/bitvec

My pleasure to create a pr about it. :)

### PR DESCRIPTION
**Pull request by bradyjoestar**
Fixes #153

### CODE REVIEW

**Comment by bradyjoestar** [comment]
: [clabot:check]

File: parity-bytes/Cargo.toml [hunk 1]
```diff
@@ -7,3 +7,9 @@ description = "byte utilities for Parity"
 license = "GPL-3.0"
 
 [dependencies]
+
+
+[features]
+default = ["std","nostd"]
+std = []
+nostd = []
```
Comment (ordian) [inline parity-bytes/Cargo.toml]: only `std` should be a feature

File: parity-bytes/src/lib.rs [hunk 2]
```diff
@@ -19,9 +19,28 @@
 //! Includes a pretty-printer for bytes, in the form of `ToPretty` and `PrettySlice`
 //! as
 
-use std::fmt;
-use std::cmp::min;
-use std::ops::{Deref, DerefMut};
+#![cfg_attr(not(feature = "std"), no_std)]
+#![cfg_attr(not(feature = "std"), feature(alloc))]
```
Comment (ordian) [inline parity-bytes/src/lib.rs]: `feature(alloc)` is stable now, no need for this line

Comment (bradyjoestar) [reply parity-bytes/src/lib.rs]: sure! :ok_hand:

File: parity-bytes/src/lib.rs [hunk 3]
```diff
@@ -19,9 +19,27 @@
...
-use std::fmt;
-use std::cmp::min;
-use std::ops::{Deref, DerefMut};
+#![cfg_attr(not(feature = "std"), no_std)]
+
+#[cfg(not(feature = "std"))]
+extern crate alloc;
+
+// Re-export libcore using an alias so that the macros can work without
+// requiring `extern crate core` downstream.
```
Comment (dvdplm) [inline parity-bytes/src/lib.rs]: Not sure I understand what you mean by "the macros" – are there macros exported from this crate?

Comment (ordian) [reply parity-bytes/src/lib.rs]: to be fair it seems to be copy-pasted across many crates: https://github.com/search?l=Rust&q=Re-export+libcore+using+an+alias+so+that+the+macros+can+work+without&type=Code

Comment (bradyjoestar) [reply parity-bytes/src/lib.rs]: I get it from the `elastic-array` and it have some marcos exported.
https://github.com/debris/elastic-array/blob/8f01572f850e6f387ad785fb1e59d599eb260ea0/src/lib.rs#L11-L28

I'm not sure whether  will be some macros exported from this crate, so I didn't delete it :smile:

File: parity-bytes/src/lib.rs (hunk 2 above)
Comment (ordian) [reply parity-bytes/src/lib.rs]: `cargo +nightly test --no-default-features` fails,
I think adding `#[macro_use]` should help

**Comment by ordian** [comment]
Would be nice adding a README with a no-std section to parity-bytes (explaining it's currently nightly-only), similar to https://github.com/paritytech/parity-wasm#no_std-crates

**Comment by bradyjoestar** [comment]
I have met the similar error  when running cargo +nightly test --no-default-features:
https://github.com/rust-lang/rust/issues/54010

#[macro_use] is added and I had forgot the `String`.

File: parity-bytes/README.md [hunk 4]
```diff
@@ -0,0 +1,11 @@
+## `no_std` crates
```
Comment (ordian) [inline parity-bytes/README.md]:
```suggestion
## `no_std` support
```

File: parity-bytes/README.md [hunk 5]
```diff
@@ -0,0 +1,11 @@
...
+
+This crate has a feature, `std`, that is enabled by default. To use this crate
+in a `no_std` context, add the following to your `Cargo.toml` (still requires allocator though):
+
+```toml
+[dependencies]
+parity-bytes = { version = "0.1.0", default-features = false }
```
Comment (ordian) [inline parity-bytes/README.md]: I think we will bump the version before release with `no-std` support.
```suggestion
parity-bytes = { version = "0.1.1", default-features = false }
```

File: parity-bytes/README.md [hunk 6]
```diff
@@ -0,0 +1,11 @@
+## `no_std` support
+
+This crate has a feature, `std`, that is enabled by default. To use this crate
+in a `no_std` context, add the following to your `Cargo.toml` (still requires allocator though):
+
+```toml
+[dependencies]
+parity-bytes = { version = "0.1.1", default-features = false }
```
Comment (dvdplm) [inline parity-bytes/README.md]: This is a nitpick but let’s do `version = “0.1”`.

File: parity-bytes/src/lib.rs (hunk 3 above)
Comment (dvdplm) [reply parity-bytes/src/lib.rs]: So if we’re not doing it for the macros, then what is the purpose of this re-export? Can we remove it?

Comment (bradyjoestar) [reply parity-bytes/src/lib.rs]: removed. :ok_hand:

File: parity-bytes/src/lib.rs [hunk 7]
```diff
@@ -19,9 +19,30 @@
...
-use std::ops::{Deref, DerefMut};
+#![cfg_attr(not(feature = "std"), no_std)]
+#![cfg_attr(not(feature = "std"), feature(alloc))]
+
+#[cfg(not(feature = "std"))]
+#[macro_use]
+extern crate alloc;
+
+#[doc(hidden)]
+extern crate core;
```
Comment (ordian) [inline parity-bytes/src/lib.rs]: https://travis-ci.org/paritytech/parity-common/jobs/533187061#L1447

Comment (bradyjoestar) [reply parity-bytes/src/lib.rs]: https://ci.appveyor.com/project/paritytech/parity-common/builds/24584892

File: parity-bytes/src/lib.rs (hunk 3 above)
Comment (bradyjoestar) [reply parity-bytes/src/lib.rs]: I delete the re-export,it show the following errors:
https://travis-ci.org/paritytech/parity-common/jobs/533187061#L1447
https://ci.appveyor.com/project/paritytech/parity-common/builds/24584892

Comment (dvdplm) [reply parity-bytes/src/lib.rs]: So I think you need to do:

```
#[cfg(feature = "std")]
extern crate core;
```
(notice no `pub`)

…and then replace `use core_::{… …` with `use core::{… …`.

The idea is that when running with `std` you need to import the `core` as a crate. In `no_std` mode `core` is part of the prelude.

Comment (bradyjoestar) [reply parity-bytes/src/lib.rs]: Thanks for your guide!

**Comment by dvdplm** [comment]
@bradyjoestar thank you for your work here! :)

**Comment by bradyjoestar** [comment]
my pleasure! :)

**Pull request merged**